Permissioned token (#292)
scale-it/algo-builder: examples/security-token/README.md

> # Security Token

Author: ok

Security tokens are permissioned tokens that carry complex compliance requirements, such as investor whitelisting and transfer restrictions. For example, token holders outside the U.S. can be prevented from transferring a token to a U.S. individual.

## Motivation

> Specifically, in Security Tokens, we want to:

Author: This will be renamed to permissioned tokens.

- Control the number of tokens minted for distribution and also allow flexibility for the administrator to issue more tokens in the future.
- Create whitelists of individuals who are approved to buy/hold the tokens and segregated sublists, as needed, to support transfer restrictions (e.g., U.S. vs international).
- Manage lock-ups as needed to ensure there are not transfers before allowed.
- Process requirements or needs that come up post-issuance, such as adding or removing investors to the whitelists or creating additional whitelists for new restrictions.
- Take confidence in the ability to remove restrictions in the future if transfers become open or unlimited.
The Security Token Template outlines a design of security tokens with emphasis for shares and equity.

## Design Overview

We want to enforce permissioned checks on all token transfer between non-reserve accounts. We will use a stateless TEAL as a clawback, which will execute all transfers. So if user A wants to transfer to user B, it will be a clawback which will transfer from account A to account B.
Steps:

1. Create assets
2. Distribute to initial users

> 3. Freeze all assets OR create the asset default-frozen (clawback can still move frozen tokens).

Author: Either freeze all assets in step 3, OR create the asset by setting default-frozen `true`. If asset is default frozen then we wouldn't need step 3.

4. Create a stateful contract with rules/permissions (`P`) and another contract: controller (`C`). P will store the C address (which can be modified).

> 5. Create a stateless teal (escrow contract `clawback-escrow`) which will check that the C and P is called as well.

Author: ok

- *permissions.py*: This is the rules smart contract. In this template we have two rules:
  - An account cannot hold more than 100 tokens
  - Both from, to accounts must be whitelisted
- *clear_state_program.py*: clear state program (returns 1)

`assets/asa.yaml` - This the the asset definition file.

### Setup

Deployment scripts are placed directly in `/scripts`:

**NOTE:** For transactions involving a multisig account (eg. token can be created and issued by a multisig), user will control one of the multisig accounts and he will receive a signed transaction i.e a signed `deployASA` tx by accounts in msig <= threshold. User can append his/her using `algob sign-multisg` command & then use `executeSignedTxFromFile` function to successfully send transaction to network (eg. deploy token).

1. *0-setup-token.js*: In this script we deploy the token (as an Algorand Standard Asset) using `deployer.deployASA(..)` function. Check note above if creator is a multisig address.

> NOTE: In the template

Author: Sorry, this will be removed.

### Execution

1. *Issuance* (`/scripts/issuance/issue.js`): Issuer can issue/mint few tokens to an account. Receiver will need to opt-in to the token first. For issuance tx, we need a transaction group of two transactions :-
a) Call to the controller smart contract - it checks if token is not killed (as issuance is not allowed if token is killed).
b) Asset transfer transaction from issuer -> receiver using clawback. `clawback-escrow.py` does the basic sanity checks for issuance_tx.
We don't need additional rule checks as issuer sets the rules himself.

2. *Update Asset Reserve* (`/scripts/issuance/update-reserve.js`): Say if reserve is a multisig address and we want to kick out one address from the multisig, then we will need to update the asset reserve address. Also, if we update the asset reserve, we will need to move all tokens from previous reserve -> new reserve. This can be done in an atomic transaction group of 2 txns:
a) Asset transfer transaction using clawback escrow where we move all tokens from previous reserve to new one.

> b) Asset Modification transfer where we update the asset reserve address to new one.

Author: Yes, added support for rekying in #296. Will update this PR after it is merged. 
However, i would like to keep this approach as well (maybe commented), because with rekeying we update an account itself (rather than a property of asset).

> b) Asset Modification transfer where we update the asset reserve address to new one.

Reviewer: OK. Then let's implement one and add a note for the alternative solution (we can even implement both).

Also, in the text above, do we need word "transfer" in "Asset Modification transfer"?

> b) Asset Modification transfer where we update the asset reserve address to new one.

Author: > OK. Then let's implement one and add a note for the alternative solution (we can even implement both).

I implemented both approaches in a separate functions - [https://github.com/scale-it/algo-builder/pull/292/files#diff-906cbbe7662016ad6886d22745796565b21389c606b587ac4f559bb352a366abR137](https://github.com/scale-it/algo-builder/pull/292/files#diff-906cbbe7662016ad6886d22745796565b21389c606b587ac4f559bb352a366abR137)

> Also, in the text above, do we need word "transfer" in "Asset Modification transfer"?

No, it was a typo. Updating to "Asset Modification transaction"

3. *OptOut* (`/scripts/issuance/opt-out.js`): User can optOut from the token if he wants to. This could be a simple asset transfer transaction from user -> creator (using user's sk) where assetamount = 0 & closeRemainderTo is asset creator. This tx will opt him out of the token and transfer all his tokens to the asset creator.
NOTE: If asset creator and reserve are different address, then we will need to transfer all tokens from creator to reserve as well (to take those tokens out of circulation/increase total supply).

4. *Asset transfer* (`scripts/transfer/transfer.js`): A non-reserve account can hold few tokens (via issuance) and may wish to transfer some tokens to another non-reserve account. This should be in complaince with rules checks (`permissions.py`). For token transfer we need a group of "atleast" 4 transactions (and more if there are more than 1 rules contracts):
a) call to controller smart contract (this ensures all rules are checked)
b) token transfer transaction from sender -> receiver using clawback (clawback contract ensures controller smart contract is called).
c) payment transaction from sender -> clawback-escrow (to cover tx fee of above transaction)

> d) call to permissions smart contract (to check rules: max token holding by receiver < 100, both from & to are whitelisted)

Author: yes

5. *Kill Token* (`scripts/permissions/kill.js`): Token manager can kill the token. If the token is killed then all issuance and token transfer transactions are rejected. User can only opt-out from the token to remove his holding and decrease his account's minimum balance.

> a) To kill the token, execute an application call tx to the controller smart contract with appArg: 'kill'. Only ASA mangager can do that. Controller updates the global state key (*is_token_killed*) to true.

Reviewer: ```suggestion

    a) To kill the token, execute an application call tx to the controller smart contract with appArg: 'kill'. Only ASA mangager can do that. Controller updates the global state key (*is_token_killed*) to true.
```

> 6. *Whitelist User* (`scripts/permissions/whitelist.js`): Permissions manager can add a new user to the whitelist which enables that user to receive or send tokens. Only permissions manager (which is stored in the global state of controller smart contract) can add a user to whitelist.

Reviewer: ```suggestion
6. *Whitelist User* (`scripts/permissions/whitelist.js`): Only a permissions manager (whose address is stored in the controller smart contract) can add a new user to the whitelist which enables that user to receive or send tokens.
```

> For executing this tx, permissions manager calls the permissions smart contract with application arg "add_whitelist", and passes the account he wishes to white list in AppAccount array. `token_id` is also passed in the ForeignAsset array to verify token. If tx is successful, then permission smart contract updates the local state of `Txn.accounts[1]` & set *whitelisted = true (Int(1))*.

Reviewer: ```suggestion
For executing this tx, permissions manager calls the permissions smart contract with application arg "add_whitelist", and passes the account he wishes to whitelist in AppAccount array. `token_id` is also passed in the ForeignAsset array to verify token. If tx is successful, then permission smart contract updates the local state of `Txn.accounts[1]` & set *whitelisted = true (Int(1))*.
```

7. *Change Permissions Manager* (`scripts/permissions/change-manager.js`): Token manager can change the permissions manager via an application call to the controller smart contract. If the parameters are correct, tx is accepted and controller smart contract updates the `permission_manager` in it's global state.

### Other Solution

> The solution in this template extends Jason's original idea - using asset clawback as escrow for security token.

Reviewer: ```suggestion
The solution in this template extends Algorand Dev Office Hour idea presented by Jason - using asset clawback as escrow for security token.
```

Reviewer: Please include a link to the webinar

Author: ok
